### ISSUE TITLE
SSL issues

### ISSUE THREAD

**Issue by MaurizioB84** [SebastianOsinski/HomebridgePluginSamsungAirConditioner#28]
Hello, after one restart of my raspberry I see this error in the log of homebridge:

Error occured: 1995928948:error:1425F102:SSL routines:ssl_choose_client_version:unsupported protocol:../deps/openssl/openssl/ssl/statem/statem_lib.c:1929:

I try to change into this file 

> /etc/ssl/openssl.cnf

the following fields:
`MinProtocol = TLSv1.2
CipherString = DEFAULT@SECLEVEL=1`

but without success... 

this is my config.json
`
        {
            "accessory": "Samsung Air Conditioner",
            "name": "Climatizzatore Cameretta",
            "ip_address": "192.168.2.90",
            "mac": "78:25:AD:12:aa:aa",
            "token": "33965903-4482-M849-N716-373832354aaa"
        }
`
and this is the logs:

> homebridge    | [10/29/2019, 12:51:09 PM] [Climatizzatore Cameretta] Connecting...
> homebridge    | [10/29/2019, 12:51:10 PM] [Climatizzatore Cameretta] Error occured: 1995928948:error:1425F102:SSL routines:ssl_choose_client_version:unsupported protocol:../deps/openssl/openssl/ssl/statem/statem_lib.c:1929:
> homebridge    | 
> homebridge    | [10/29/2019, 12:51:10 PM] [Climatizzatore Cameretta] Connection closed because error occured
> homebridge    | [10/29/2019, 12:51:10 PM] [Climatizzatore Cameretta] Trying to reconnect in 5s...
>

**Comment by MaurizioB84** [comment]
maybe it is the new version of node, in the container I have 12.13.0 and I think required TLS1.2 can you please verify if is it possible to upgrade the tls version?

can you also provide me a workaround?

**Comment by SebastianOsinski** [comment]
Hi @MaurizioB84, sorry for the big delay. It's hard for me to suggest any solution here other than checking other Node version. I'm not JS developer, this plugin is my first Node project ;).

**Comment by MaurizioB84** [comment]
in the mind-time I found the issues, is related to tls 1.2, you need to upgrade node version with pls 1.2, is it possible?

**Comment by SebastianOsinski** [comment]
@MaurizioB84 Hi, could you try other NodeJS version? I'm currently using 10.16.3 without any problems. TBH I don't have time right now to look into this issue. Any Pull Requests resolving it are highly welcome though :)

**Comment by init4** [comment]
I had this problem too. It's because node isn't allowing TLS 1.0 (which is what the Samsung AC supports, or SSLv3).

To fix I changed the homebridge helper script (e.g. /usr/local/bin/homebridge) and add "--tls-min-v1.0 ".

Somewhere along the lines the node default changed to stop TLS 1.0, it certainly wasn't a problem until I updated recently.

**Comment by peros550** [comment]
@init4 I'm not an experienced user. 
I have found the helper script, but have no clue where to put "--tls-min-v1.0 " 

Could you please provide some help?

Many thanks,

**Comment by Anzic23** [comment]
+1

**Comment by init4** [comment]
@peros550 @Anzic23 Find the main homebridge executable (which is just a shell script), for me it was /usr/local/bin/homebridge - edit this file and change the first line.

The first line will look something like this:
#!/usr/bin/env node

Change the first line to look like this:
#!/usr/bin/env node --tls-min-v1.0

Then you should be good to go.

**Comment by peros550** [comment]
@init4 many thanks for your message. 
I tried the proposed change but I still get the same error message 

Any other ideas?

```

[6/19/2020, 19:42:04] Error: ca md too weak
    at Object.createSecureContext (_tls_common.js:273:19)
    at Object.connect (_tls_wrap.js:1560:48)
    at AirConditionerApi.connect (/usr/local/lib/node_modules/homebridge-plugin-samsung-air-conditioner/lib/air-conditioner-api.js:49:27)
    at AirConditioner.getServices (/usr/local/lib/node_modules/homebridge-plugin-samsung-air-conditioner/lib/air-conditioner.js:38:18)
    at Server.createHAPAccessory (/usr/local/lib/node_modules/homebridge/src/server.ts:441:41)
    at /usr/local/lib/node_modules/homebridge/src/server.ts:359:30
    at Array.forEach (<anonymous>)
    at Server._loadAccessories (/usr/local/lib/node_modules/homebridge/src/server.ts:329:29)
    at Server.start (/usr/local/lib/node_modules/homebridge/src/server.ts:155:12)
    at cli (/usr/local/lib/node_modules/homebridge/src/cli.ts:80:10)
```

**Comment by Anzic23** [comment]
changed the file. the error is still there.
Node.js | v12.18.1
my error is - `Error occured: 139995822624072:error:1425F102:SSL routines:ssl_choose_client_version:unsupported protocol:../deps/openssl/openssl/ssl/statem/statem_lib.c:1942:`

**Comment by peros550** [comment]
Is there a way to use this plugin with the current state of Homebridge + node ? I know @SebastianOsinski  plans to update it for current node version but I was wondering if in the meanwhile there could be any workaround?

I tried to setup a new raspberry running an older version of Homebridge & Node v8 but I wasn't successful at all. First, I wasn't able to install node v8.  Then, I found NVE and was able to pick latest v8 node. But all put together I wasn't successful, some sort of  dependency incompatibility error came up. 

If anybody could provide some guidance it would be very appreciated and helpful!

**Comment by djsomi** [comment]
Same here,

Even I cannot start homebridge with edited line: #!/usr/bin/env node --tls-min-v1.0
Drops:  /usr/bin/env: ‘node --tls-min-v1.0’: No such file or directory

I have to roll back a bakcup, to get it working again. (it was on node v11.15.0)
Anyone get this successfully working with node v12+?

Thanks
Zoli

**Comment by init4** [comment]
The important part of the change is the argument being passed to node. I have no idea why the env command can't find your node binary, but if you add the suggested argument to your hashbang (the first line) it should work.

So if your first line is something like:

#!/usr/local/bin/node

Then it would become:

#!/usr/local/bin/node --tls-min-v1.0

Same goes for wherever your node binary is installed, whether its /usr/bin, /usr/local/bin, /opt/local/bin, etc etc.

**Comment by Anzic23** [comment]
in my case this help
NODE_OPTIONS set to --tls-min-v1.0
file /usr/local/bin/homebridge unchanged (first line "#!/usr/bin/env node")
![image](https://user-images.githubusercontent.com/37106885/87222206-a03cff80-c37a-11ea-9ab9-4180ba04e6bd.png)

**Comment by peros550** [comment]
@Anzic23 could you please share your setup ? Homebridge / node versions and any other information you may have? 
I have tried this option to a newly installed homebridge (v1.1.1), node version 12.18.2 and I still get the same error.

In my case, homebridge has no other accessories in json but the AC.

**Comment by Anzic23** [comment]
@peros550 sure:

Node.js | v12.18.2
```
homebridge (v1.1.1)
"accessories": [
        {
            "accessory": "Samsung Air Conditioner",
            "name": "Air Conditioner",
            "ip_address": "192.168.1.71",
            "mac": "BC:8C:33:33:33:33",
            "token": "6df****50f"
        }
```

**Comment by peros550** [comment]
@Anzic23 Many thanks for your support. We have the same setup. :(

Did you also try the command `homebridge-samsung-ac-get-token <your ac's ip address>` right after installing the plugin?

I installed the plugin with this command : `sudo npm install -g homebridge-plugin-samsung-air-conditioner --unsafe-perm` but I above command gives again the known Error: ca md too weak. 

I have kept the token from a previous installation.

**Comment by Anzic23** [comment]
@peros550 I didn't run the token search, I still have it from the last installation.

**Comment by peros550** [comment]
@Anzic23 many thanks for your help! I did everything you mentioned and I also found very helpful this [topic](https://github.com/cicciovo/homebridge-samsung-airconditioner/issues/17#issuecomment-616004492)

**Comment by djsomi** [comment]
> #!/usr/local/bin/node --tls-min-v1.0

Thanks, its working fine!

**Comment by djsomi** [comment]
> in my case this help
> NODE_OPTIONS set to --tls-min-v1.0
> file /usr/local/bin/homebridge unchanged (first line "#!/usr/bin/env node")
> ![image](https://user-images.githubusercontent.com/37106885/87222206-a03cff80-c37a-11ea-9ab9-4180ba04e6bd.png)

Thanks this is also a working solution, thank you all!

**Comment by Jens-Wymeersch** [comment]
> @peros550 @Anzic23 Find the main homebridge executable (which is just a shell script), for me it was /usr/local/bin/homebridge - edit this file and change the first line.
> 
> The first line will look something like this:
> #!/usr/bin/env node
> 
> Change the first line to look like this:
> #!/usr/bin/env node --tls-min-v1.0
> 
> Then you should be good to go.

Hello,
I'm looking for this main executable but don't find it. Can you please indicate what i'm looking for ? I found the Homebridge directory at /var/lib/homebridge but the /usr/local/bin/ is empty.

Thank you so much,
Jens Wymeersch

**Comment by lidans5** [comment]
for next generations to be:
followed that instruction and stopped getting the above error: (still being refused, but at least get the listening mode fine)
[solution](https://askubuntu.com/questions/1231799/certificate-error-after-upgrade-to-20-04)

**Comment by lidans5** [comment]
I'll appreciate any idea regarding next message I get:
Failed to connect to 192.168.0.139 port 8888: Connection refused

**Comment by SanderBaron** [comment]
I'm also one of the unlucky guys stuck with 4 older models Samsung AC's... 😞 and I'm desperately trying to get this plugin to work for now and the future! I'm on HB 1.6.0, running on a MacMini. Node js. v16.14.0.
I tried the "--tls-min-v1.0" option command in HB and this fixed the problem partly. Plugin is running again and AC's are showing up in the Home app. Only downside is that I can't control the AC, it switched straight back to "off" when I try to change setting in the Home app.... 

Would be so cool if some one could fix this. I see a lot of forks, but is any of them working perfect?  I'm not a code guy myself but I'm good in following instructions. Also willing to pay or make a donation for a good fix. 

Thanks!

**Comment by SanderBaron** [comment]
> Hello, @Anzic23, @SebastianOsinski thank you for your suggestion. It worked like a charm. Now I'm stuck with the fact that I'm not able to adjust the temperature. Any ideas ?
> 
> Thank you so much, Jens Wymeersch

Hi @Jens-Wymeersch ,
I have the same problem. It works but I'm not able to adjust anything. Did you managed to solve this? Thanks!

**Comment by Wrotaz** [comment]
Probably not the solution you are looking for, but … I started using the plug-in years ago. It worked like a charm, but after a few updates of homebridge and nodejs it stopped working. So, (after many unsuccessful attempts) rolled back the updates, moved the hombridge with this single plugin to another instance (virtual machine) and stopped updating. And it’s still working. But the nodejs is at v10.16 and homebridge at v0.4.46. Yes. They are this old :-).

**Comment by Jens-Wymeersch** [comment]
Gents,

This was always a temp solution for me till the addon in home assistant would work which it does for years. 

BR, 
Jens

**Comment by djsomi** [comment]
> Gents,
> 
> This was always a temp solution for me till the addon in home assistant would work which it does for years.
> 
> BR, Jens

Jens, can you please share which HASS addon are you using?

Anyway this plugin still working for me as mentioned above, the only showstopper is the nodejs now, have to stay on V16, above that plugin fails.

**Comment by Jens-Wymeersch** [comment]
Here you go : https://github.com/atxbyea/samsungrac

**Comment by lidans5** [comment]
Well
The problem for me always was how do you get the token

**Comment by SanderBaron** [comment]
Luckily I have backed up the tokens! Will try HA now![Sent from my iPhone]Op 12 mrt. 2023 om 15:25 heeft lidans5 ***@***.***> het volgende geschreven:﻿
Well
The problem for me always was how do you get the token

**Comment by SanderBaron** [comment]
> > Gents,
> > This was always a temp solution for me till the addon in home assistant would work which it does for years.
> > BR, Jens
> 
> Jens, can you please share which HASS addon are you using?
> 
> Anyway this plugin still working for me as mentioned above, the only showstopper is the nodejs now, have to stay on V16, above that plugin fails.

Been playing around with HA this afternoon... Have to admit that this maybe is a level too high for this boomer... 😞

**Comment by lidans5** [comment]
Same here 😭
I need a detailed guide

**Comment by SanderBaron** [comment]
@lidans5 Yep, eventually I will get it to work but the real problem starts when I have to troubleshoot after 6 months... I literally won't remember anything and I will have to start with a clean install again... Happend to me to many times... 😞 

So, if anyone is willing to help me to fix the plugin to modern standards I would be very happy! Willing to pay some or make a donation... Thanks!

**Comment by huniak** [comment]
I sudgest byť something like this
https://www.google.com/search?q=ir+blaster&client=ms-android-oppo-rvo3&prmd=sivn&source=lnms&tbm=shop&sa=X&ved=2ahUKEwj5xrjr9tb9AhXiS_EDHSsjAccQ_AUoAXoECAIQAQ&biw=360&bih=665&dpr=3
And control your climatization Via application

Dňa ne 12. 3. 2023, 12:33 SanderBaron ***@***.***> napísal(a):

**Comment by SanderBaron** [comment]
> I sudgest byť something like this https://www.google.com/search?q=ir+blaster&client=ms-android-oppo-rvo3&prmd=sivn&source=lnms&tbm=shop&sa=X&ved=2ahUKEwj5xrjr9tb9AhXiS_EDHSsjAccQ_AUoAXoECAIQAQ&biw=360&bih=665&dpr=3 And control your climatization Via application Dňa ne 12. 3. 2023, 12:33 SanderBaron ***@***.***> napísal(a):
> […](#)

Haha, just because I'm a boomer, that doesn't mean I can't have standards! 😃

**Comment by SanderBaron** [comment]
> Hi guys! I’m the maintainer of this repository. I have had couple attempts to start fixing this, but unfortunately I really dislike working with JavaScript 😅. I might come back to it in near future (probably before July/August) to make it work again. However I can’t promise anything as I myself migrated to HomeAssistant some time ago. I have yet to decide if I will use HA integration or try to revive this one.
> 
> If anyone feels up to the challenge to fix this, they’re more than welcome, that’s what open source is all about ;)

That would be great @SebastianOsinski ! Do you use HA now together with Apple Homekit?

**Comment by djsomi** [comment]
Guys, 

Just a simple tutorial: 

- Install homebridge in a docker (fast and easy for testing): [https://github.com/homebridge/homebridge/wiki/Install-Homebridge-on-Docker](url)
- Once homebridge is up, do the initial setup than go to homebridge settings, put this into node options:  --tls-min-v1.0 than restart it.
- Install homebridge-plugin-samsung-air-conditioner v2.0.0, edit config file as needed, than restart it. You will see plugin crashes homebridge.
- Issue this command on host system: docker exec [YOURHBDOCKERNAME] hb-service update-node 16.19.1 --allow-root than restart homebridge again.
- Check nodejs version: docker exec [YOURHBDOCKERNAME] node -v

Plugin now works fine. 

BR
Zoltan

**Comment by SanderBaron** [comment]
> Guys,
> 
> Just a simple tutorial:
> 
> * Install homebridge in a docker (fast and easy for testing): [https://github.com/homebridge/homebridge/wiki/Install-Homebridge-on-Docker](url)
> * Once homebridge is up, do the initial setup than go to homebridge settings, put this into node options:  --tls-min-v1.0 than restart it.
> * Install homebridge-plugin-samsung-air-conditioner v2.0.0, edit config file as needed, than restart it. You will see plugin crashes homebridge.
> * Issue this command on host system: docker exec [YOURHBDOCKERNAME] hb-service update-node 16.19.1 --allow-root than restart homebridge again.
> * Check nodejs version: docker exec [YOURHBDOCKERNAME] node -v
> 
> Plugin now works fine.
> 
> BR Zoltan

Hi Zoltan,
Thanks for the helping hand!!
I'm running HB on a Mac and I read it doesn't work for Docker on Mac. 
But what exactly does this command? Can I use it as a Terminal command on my Mac? And how to reverse it?
I'm running HomeBridge solely as a test now for the Samsung Plugin. Also running Hoobs on the same Mac for my other plugins.

Thanks, Sander

**Comment by init4** [comment]
@SanderBaron 

I run HB on a Mac too. I'm using node 14.17.4, homebridge 1.4.0 & samsung-ac plugin version 2.0 (according to npms - doesn't seem to match a version here on github). I have several Samsung ACs and they all work without issues as long as you are setting --tls-min-v1.0.

The trickiest part of setting the Samsung AC's up is to collect the tokens using 'homebridge-samsung-ac-get-token <ip>'. This will involve running the command, and then going to press the power button on the AC remote. The token should then appear in your 'homebridge-samsung-ac-get-token <ip>' command output. Once you have a token, keep it safe somewhere as it will save a lot of hassle in future. If you have doubts, generate the tokens again.

My only other advice is to set static IPs for all the AC's using your DHCP server. I suspect that most home automation device problems generally come down to patchy wifi and devices that are changing IPs.

Once you have homebridge installed with the plugin, you just have to define the appropriate homebridge config. Something like:

{
    "accessory": "Samsung Air Conditioner",
    "name": "Study",
    "ip_address": "10.0.0.42",
    "mac": "54:88:0E:65:A6:DE",
    "token": "5c7441ba-765a-41e1-8da1-c3de1dbc3dea"
}

Make sure you are matching (& verifying) the correct AC, MAC, IP & token. The homebridge logs should give you a clue if something isn't configured correctly.

**Comment by SebastianOsinski** [comment]
>Do you use HA now together with Apple Homekit?

Yes, I do, works flawlessly.

**Comment by SanderBaron** [comment]
> > Do you use HA now together with Apple Homekit?
> 
> Yes, I do, works flawlessly.

That's good to hear @SebastianOsinski ! I'm still learning... I have HA running now on my Mac, but I can't figure out how to get exposure to HomeKit and how and where to install the plugin for Samsung AC. 
Can you tell me what you are running? Thanks!

**Comment by SanderBaron** [comment]
> Guys,
> 
> Just a simple tutorial:
> 
> * Install homebridge in a docker (fast and easy for testing): [https://github.com/homebridge/homebridge/wiki/Install-Homebridge-on-Docker](url)
> * Once homebridge is up, do the initial setup than go to homebridge settings, put this into node options:  --tls-min-v1.0 than restart it.
> * Install homebridge-plugin-samsung-air-conditioner v2.0.0, edit config file as needed, than restart it. You will see plugin crashes homebridge.
> * Issue this command on host system: docker exec [YOURHBDOCKERNAME] hb-service update-node 16.19.1 --allow-root than restart homebridge again.
> * Check nodejs version: docker exec [YOURHBDOCKERNAME] node -v
> 
> Plugin now works fine.
> 
> BR Zoltan

@djsomi I installed node 16.19.1 as per your instructions. Still same problem. AC is exposed to HomeKit, but you can't change settings. The moment you change settings on your phone, it immediately switches back to "off". See log:

[3/19/2023, 10:13:37 PM] [Sam] Setting target state: 0
[3/19/2023, 10:13:37 PM] [Sam] Target state set
[3/19/2023, 10:13:37 PM] [Sam] Setting active: 1
[3/19/2023, 10:13:37 PM] [Sam] Active set

Any other suggestions to try? 

Thanks, Sander

**Comment by SanderBaron** [comment]
Hi, anyone in here who has experience with AI to review this code and bring it up to date with latest node? 
If no one has time, why not let AI work for it? 😃

**Comment by SebastianOsinski** [comment]
Hi everyone. I just released 3.0.0 version of the plugin. It should resolve all SSL issues on new Node versions. I tested it with current homebridge docker image and it works great

**Comment by djsomi** [comment]
> Hi everyone. I just released 3.0.0 version of the plugin. It should resolve all SSL issues on new Node versions. I tested it with current homebridge docker image and it works great

I can also confirm, in production system, new plugin works with v18.16.0!

Thank you very much!

**Comment by SanderBaron** [comment]
Can confirm that is does work!!

Thanks a lot Sebastian!!
